File: android/app/src/main/java/com/nauhaan/skycast/ui/forecast/ForecastViewModel.kt
package com.nauhaan.skycast.ui.forecast

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.nauhaan.skycast.core.common.AppError
import com.nauhaan.skycast.domain.model.Forecast
import com.nauhaan.skycast.domain.model.SavedLocation
import com.nauhaan.skycast.domain.model.UserPreferences
import com.nauhaan.skycast.domain.repository.DataState
import com.nauhaan.skycast.domain.repository.LocationRepository
import com.nauhaan.skycast.domain.repository.SettingsRepository
import com.nauhaan.skycast.domain.repository.WeatherRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.flatMapLatest
import kotlinx.coroutines.flow.flowOf
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * State for the Forecast tab.
 *
 * Same derived-property approach as `TodayUiState`: the offline-first display rules live on the
 * state, so the composable never combines flags itself.
 */
data class ForecastUiState(
    val location: SavedLocation? = null,
    val forecast: Forecast? = null,
    val preferences: UserPreferences = UserPreferences(),
    val isLoading: Boolean = false,
    val isRefreshing: Boolean = false,
    val isStale: Boolean = false,
    val error: AppError? = null,
    val hasNoLocation: Boolean = false,
) {
    val showsFullScreenLoader: Boolean get() = isLoading && forecast == null
    val showsFullScreenError: Boolean get() = error != null && forecast == null && !hasNoLocation
    val showsEmptyState: Boolean get() = hasNoLocation && !isLoading
    val showsContent: Boolean get() = forecast != null && forecast.days.isNotEmpty()
    val showsStaleBanner: Boolean get() = showsContent && (error != null || isStale)
}

/**
 * View model for the Forecast tab.
 *
 * Mirrors `TodayViewModel`: resubscribes when the primary location changes, and combines the
 * unit preferences in so switching °C→°F re-renders from cache with no network call.
 */
@HiltViewModel
class ForecastViewModel
@Inject
constructor(
    private val locationRepository: LocationRepository,
    private val weatherRepository: WeatherRepository,
    settingsRepository: SettingsRepository,
) : ViewModel() {
    private val manualRefreshInFlight = MutableStateFlow(false)

    /**
     * The error from the most recent **manual** refresh. See `TodayViewModel` for why this is
     * kept apart from the observed `DataState.error`: a failed pull-to-refresh does not
     * necessarily make the stream re-emit, so without this the failure is silent.
     */
    private val manualRefreshError = MutableStateFlow<AppError?>(null)

    @OptIn(ExperimentalCoroutinesApi::class)
    val uiState: StateFlow<ForecastUiState> =
        locationRepository
            .observePrimaryLocation()
            // flatMapLatest, not combine: a changed primary location must cancel the previous
            // location's forecast subscription rather than merge with it.
            .flatMapLatest { location ->
                if (location == null) {
                    flowOf(DataState<Forecast>() to null)
                } else {
                    weatherRepository
                        .observeForecast(location)
                        .combine(flowOf(location)) { state, loc -> state to loc }
                }
            }
            .combine(settingsRepository.observePreferences()) { (state, location), preferences ->
                Triple(state, location, preferences)
            }
            .combine(manualRefreshInFlight) { triple, manualRefresh -> triple to manualRefresh }
            .combine(manualRefreshError) { (triple, manualRefresh), refreshError ->
                val (state, location, preferences) = triple
                ForecastUiState(
                    location = location,
                    forecast = state.data,
                    preferences = preferences,
                    isLoading = state.isLoading,
                    isRefreshing = state.isRefreshing || manualRefresh,
                    isStale = state.isStale,
                    error = state.error ?: refreshError,
                    hasNoLocation = location == null,
                )
            }
            .stateIn(
                scope = viewModelScope,
                started = SharingStarted.WhileSubscribed(SUBSCRIPTION_TIMEOUT_MILLIS),
                initialValue = ForecastUiState(isLoading = true),
            )

    fun refresh() {
        val location = uiState.value.location ?: return
        viewModelScope.launch {
            manualRefreshInFlight.value = true
            manualRefreshError.value = weatherRepository.refresh(location)
            manualRefreshInFlight.value = false
        }
    }

    private companion object {
        const val SUBSCRIPTION_TIMEOUT_MILLIS = 5_000L
    }
}

File: ios/SkyCast/Features/Forecast/ForecastViewModel.swift
import Foundation
import Observation

/// Everything the Forecast screen renders.
///
/// Same derived-property approach as ``TodayUiState``: the offline-first display rules live on
/// the state, so the view never combines flags itself.
struct ForecastUiState: Equatable {
    var location: SavedLocation?
    var forecast: Forecast?
    var preferences: UserPreferences = .init()
    var isLoading = false
    var isRefreshing = false
    var isStale = false
    var error: AppError?
    var hasNoLocation = false
    var isBannerDismissed = false

    var showsFullScreenLoader: Bool {
        isLoading && forecast == nil
    }

    var showsFullScreenError: Bool {
        error != nil && forecast == nil && !hasNoLocation
    }

    var showsEmptyState: Bool {
        hasNoLocation && !isLoading
    }

    var showsContent: Bool {
        forecast?.days.isEmpty == false
    }

    var showsStaleBanner: Bool {
        showsContent && !isBannerDismissed && (error != nil || isStale)
    }

    var staleBannerMessage: String {
        error?.message ?? "Showing saved data, it may be out of date."
    }
}

/// View model for the Forecast tab.
///
/// Mirrors ``TodayViewModel`` deliberately, including the single held `observationTask`: starting
/// a second stream without cancelling the first would leave two writers racing on `state`.
@MainActor
@Observable
final class ForecastViewModel {
    private(set) var state = ForecastUiState(isLoading: true)

    private let weatherRepository: any WeatherRepository
    private let locationRepository: any LocationRepository
    private let settingsStore: SettingsStore

    private var observationTask: Task<Void, Never>?

    init(
        weatherRepository: any WeatherRepository,
        locationRepository: any LocationRepository,
        settingsStore: SettingsStore
    ) {
        self.weatherRepository = weatherRepository
        self.locationRepository = locationRepository
        self.settingsStore = settingsStore
    }

    func start() {
        guard observationTask == nil else { return }
        observationTask = Task { await observe() }
    }

    func stop() {
        observationTask?.cancel()
        observationTask = nil
    }

    func refresh() async {
        guard let location = state.location else { return }

        state.isRefreshing = true
        state.isBannerDismissed = false

        let error = await weatherRepository.refresh(location)

        state.isRefreshing = false
        if error == nil {
            await observeForecast(for: location)
        } else {
            state.error = error
        }
    }

    func dismissBanner() {
        state.isBannerDismissed = true
    }

    /// Re-reads the primary location and its forecast.
    ///
    /// Called when the tab reappears, because the user may have changed the primary place on the
    /// Locations tab while this screen was off-screen. Android gets this for free from Room's
    /// `Flow`; SwiftData has no equivalent outside a `View`, so the reload is explicit.
    func reload() async {
        stop()
        state.isLoading = state.forecast == nil
        start()
    }

    // MARK: - Internals

    private func observe() async {
        state.preferences = settingsStore.preferences

        do {
            guard let location = try await locationRepository.primaryLocation() else {
                state.isLoading = false
                state.hasNoLocation = true
                return
            }
            state.location = location
            state.hasNoLocation = false
            await observeForecast(for: location)
        } catch {
            state.isLoading = false
            state.error = AppError.from(error)
        }
    }

    private func observeForecast(for location: SavedLocation) async {
        for await dataState in weatherRepository.forecast(for: location) {
            if Task.isCancelled {
                return
            }
            apply(dataState, location: location)
        }
    }

    private func apply(_ dataState: DataState<Forecast>, location: SavedLocation) {
        state.location = location
        // Keep the previous forecast on failure, this is the visible half of the offline-first
        // promise, and the reason `DataState` carries data and error together.
        state.forecast = dataState.data ?? state.forecast
        state.isLoading = dataState.isLoading
        state.isRefreshing = dataState.isRefreshing
        state.isStale = dataState.isStale
        state.error = dataState.error
        state.preferences = settingsStore.preferences
    }
}
